### PR TITLE
Add Quick-Access to favorite folder in left sidepanel in files-app #9720

### PR DESCRIPTION
**Pull request by skjnldsv**
Somehow broke #9720 (sorry @newhinton) :(

Please review @nextcloud/designers 
Already a :+1: from me!

### CODE REVIEW

**Review by juliusknorr — DISMISSED** [review]
- [ ] Toggling the quick access list is not possible
- [x] I think we should hide it by default, otherwise some users who already have a bunch of favorites could end up with a long list of folders there and the other sections would be hidden on the first run
- [ ] Adding a folder to the favorites will not update the list in the sidebar
- [x] Clicking the folder flashes a second indicator, which looks a bit weird.
image

- [x] There is also a JS error when clicking a folder in the sidebar:
``` js
Uncaught Error: Syntax error, unrecognized expression: li[data-id=http://localhost:8140/index.php/apps/files/?dir=/4 ]
    at Function.ga.error (core.js?v=d502b209-3:2)
    at ga.tokenize (core.js?v=d502b209-3:2)
    at Function.ga [as find] (core.js?v=d502b209-3:2)
    at a.fn.init.find (core.js?v=d502b209-3:2)
    at a.fn.init.a.fn.find (core.js?v=d502b209-3:7)
    at Navigation.setActiveItem (merged-index.js?v=d502b209-3:11396)
    at Navigation._onClickItem (merged-index.js?v=d502b209-3:11421)
    at HTMLDivElement.dispatch (core.js?v=d502b209-3:3)
    at HTMLDivElement.r.handle (core.js?v=d502b209-3:3)
```

**Comment by juliushaertl** [comment]
@newhinton Besides those comments, I really like it. 

@skjnldsv @newhinton Drag-and-drop is not included in the PR, right?

**Comment by skjnldsv** [comment]
>> Toggling the quick access list is not possible
>> Clicking the folder flashes a second indicator, which looks a bit weird.
>
> I cant reproduce those

I can, I'll fix it :)

> At the moment it is only disabled. The code is still there, because i thougt that the "how to sort"-discussion wasn't over yet. I have the idea of an dropdown menu, which allows changing of the sorting-order, but i wanted a statement for that because that would introduce a dropdown-menu to the ui, which needs to be discussed. If you want to take a peak, look at newhinton:WithSettingsDropdown. I have tested it there.

@newhinton you can push to this pr!
Nonetheless, let's not add a dropdown, @jancborchardt is not fully on board for now, so let’s just sort them by last modified up top. :)

**Comment by juliushaertl** [comment]
@newhinton Chrome 67

>>Adding a folder to the favorites will not update the list in the sidebar
> Do you mean the ordering is not updated?

The folder that is added to the favorites is missing in the sidebar until i reload the whole page.

**Comment by newhinton** [comment]
@juliushaertl Sorry, i was confused there, drag&drop is not implemented at all. What i meant was the custom order, which is ordered by drag&drop :D

@skjnldsv thanks for fixing it!

File: apps/files/lib/Controller/ViewController.php [hunk 1]
```diff
@@ -159,28 +160,76 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
...
+		}
+
+		$favoritesSublistArray = Array();
+
+		$navBarPositionPosition = 6;
+		$currentCount = 0;
+		foreach ($favElements['folders'] as $elem) {
+
+			$id = substr($elem, strrpos($elem, '/') + 1, strlen($elem));
+			$link = $this->urlGenerator->linkToRouteAbsolute('files.view.index', ['dir' => $elem]);
```
Comment (skjnldsv) [inline apps/files/lib/Controller/ViewController.php]: Somehow the link generated here does not redirect to the proper view in fies :/
@juliushaertl  any clue?

Comment (skjnldsv) [reply apps/files/lib/Controller/ViewController.php]: clicking this redirect from `/apps/files/?dir=/Test`
to `/apps/files/dir=/&view=Test`

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
Unit test failure:

```
1) OCA\Files\Tests\Controller\ViewControllerTest::testIndexWithRegularBrowser
Your test case is not allowed to access the database.
```

**Comment by skjnldsv** [comment]
> Unit test failure:

Any clue? I'm on the jsunit atm :)

File: apps/files/lib/Controller/ViewController.php [hunk 2]
```diff
@@ -159,28 +160,77 @@ public function index($dir = '', $view = '', $fileid = null, $fileNotFound = fal
...
 		$storageInfo = $this->getStorageInfo();
 
+		$user = $this->userSession->getUser()->getUID();
+
+		//Get Favorite-Folder
+		$tagger = \OC::$server->getTagManager();
+		$helper = new \OCA\Files\Activity\Helper($tagger);
+
+		try {
+			$favElements = $helper->getFavoriteFilePaths($this->userSession->getUser()->getUID());
```
Comment (MorrisJobke) [inline apps/files/lib/Controller/ViewController.php]: This one is likely to cause the DB used exception. This one is called in a test that has no DB but actually needs one.

Comment (rullzer) [reply apps/files/lib/Controller/ViewController.php]: I think I fixed the tests

**Review by juliusknorr — APPROVED** [review]
Works great for me now :+1: 

Good to go, if the jsunit tests are fixed. :wink:

**Comment by skjnldsv** [comment]
All fixed, let's wait for final tests
https://drone.nextcloud.com/nextcloud/server/8883

**Comment by rullzer** [comment]
phpunit test broken! fix pushed

**Comment by newhinton** [comment]
@skjnldsv I see you have removed the _setOnDrag function, i get that, but why do you keep all the other stuff, which is now dead code?

**Comment by skjnldsv** [comment]
@newhinton I removed it because the drag wasn't working properly :)
I thought I cleaned all the related stuff, could you point out where there are still stuff left please?

Thanks for your amazing work, this pr is a very nice addition to nc14!!

**Comment by newhinton** [comment]
@skjnldsv I would prefer to fix it instead of removing it, what do you say?

**Comment by newhinton** [comment]
In that case i meant the sortable handler (for custom sorting). While we are not using it for this feature, the handling of different sorting strategies is fully working in the current files-app, and the only thing currently breaking sortablelist is the missing handler.
 When we would readd the handler, it would work again (after fixing your mentioned bugs).

This enables us to use this feature later, when (and if) we found a solution for the ui-problem (look at [newhinton:WithSettingsDropdown](https://github.com/newhinton/server/tree/WithSettingsDropdown) , i have experimented there) without redeveloping everything.
Also, slight changes could make the draggablething dynamic for multiple lists, like the upcoming pr for the shared entries, as an example.

If you want to remove it, (i can remove it, if you whish) you need to delete the api-routes, you need to remove the unnessessary sorting-variable in navigation.js, edit the quicksort-helper function to only return the alphabetical sorting value and edit the php-code to dump the initial-sorting.

I believe my solution for sorting is quiet dynamic there, and while not perfect, provides a lot of possibilities and i would keep it, but in the end, it's your decision ;)

**Comment by skjnldsv** [comment]
No, we don't want to add a menu of any sort in the end. Let keep things simple, the ui already start to be a bit overcrowded :)

Please, go ahead for the cleanup, I'm on another issue ;)
Thanks a lot!

**Comment by newhinton** [comment]
@skjnldsv How do i push the cleanup? Should i open a new PR?

**Comment by skjnldsv** [comment]
Yes, on nextcloud/server please :)

**Pull request merged**